### ISSUE TITLE
Channel design for parallel jobs for 

### ISSUE THREAD

**Issue by liuyangzzu** [nextflow-io/nextflow#2901]
I have three process: A, B, and C.

A: do basecalling on input data, the input can be more than 1 input files, such as 10 parallel jobs/batches. The output is basecalled results, will be used by both B and C.

B: take each of A’s output and do alignment, the output will be used by C.

C: need take each input batch’s both A’s output and the corresponding B’s output as input, and perform processing.

I have no idea of how to build the channel input for process C. Here, I need each batch of C take the exactly same batch's output for A and B.

Thank you very much for your help!

**Comment by liuyangzzu** [comment]
For example, if the input is 10 files, A, B, C will all have 10 jobs.

A1 -> B1
...
A3->B3

I want the process C take A1, B1 output, or Ak,Bk's output as the input. The batch number for A and B need correlated.

**Comment by bentsherman** [comment]
The way I typically do this is to turn each input file into a tuple of the form `[id, filename]`, where `id` is usually a sample id embedded in the name of the input file. If you make processes A and B output the same structure, you can then use the [groupTuple](https://www.nextflow.io/docs/latest/operator.html#grouptuple) operator to combine the outputs from A and B into a single channel where the outputs are paired correctly. I think you could also use [join](https://www.nextflow.io/docs/latest/operator.html#join) which is simpler.

**Comment by liuyangzzu** [comment]
Thank you bentsherman!
Can this method send the matched channel as soon as some id complete?
For example, if id3's A and B output is ready, but others (i.e., id1, id2 is not ready). Can id3's output be sent to C to run?

**Comment by bentsherman** [comment]
Yes, each output pair will be emitted as soon as you are ready. If you use groupTuple, you need to specify `size: 2` so that it knows to emit when two items are matched.

**Comment by liuyangzzu** [comment]
Works very well!
